### PR TITLE
[CodeCamp] Add doc of yolov6_description.md

### PR DESCRIPTION
**Pull request by satuoqaq**
## Motivation

Add doc of yolov6_description.md

## Modification

Add description of Label Assigment and Loss function

### CODE REVIEW

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md [hunk 1]
```diff
@@ -65,26 +65,332 @@ YOLOv6 目标检测算法中使用的数据增强与 YOLOv5 基本一致，唯
...
 
+YOLOv6 采用的标签匹配策略与 [TOOD](https://arxiv.org/abs/2108.07755)
+相同, 前 4 个 epoch 采用 `ATSSAssigner` 作为标签匹配策略的 `warm-up` ,
+后续使用 `TaskAlignedAssigner` 算法选择正负样本, 基于官方开源代码,  `MMYOLO` 中也对两个 assigner 算法进行了优化, 改进为 `Batch` 维度进行计算，
+能够一定程度的加快速度。 下面会对每个部分进行详细说明。
+
 #### 1.3.1 Anchor 设置
 
-YOLOv6 采用与 YOLOX 一样的 Anchor-free 无锚范式，省略的了聚类和繁琐的Anchor超参设定，泛化能力强，解码逻辑简单。在训练的过程中会根据 feature size 去自动生成先验框。
+YOLOv6 采用与 YOLOX 一样的 Anchor-free 无锚范式，省略的了聚类和繁琐的 Anchor 超参设定，泛化能力强，解码逻辑简单。在训练的过程中会根据 feature size 去自动生成先验框。
```
Comment (hhaAndroid) [inline docs/zh_cn/algorithm_descriptions/yolov6_description.md:75]:
```suggestion
YOLOv6 采用与 YOLOX 一样的 Anchor-free 无锚范式，省略了聚类和繁琐的 Anchor 超参设定，泛化能力强，解码逻辑简单。在训练的过程中会根据 feature size 去自动生成先验框。
```

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md [hunk 2]
```diff
@@ -65,26 +65,332 @@ YOLOv6 目标检测算法中使用的数据增强与 YOLOv5 基本一致，唯
...
+    y2 = points[..., 1] + distance[..., 3]
+
+    bboxes = torch.stack([x1, y1, x2, y2], -1)
+
+    return bboxes
+```
 
 #### 1.3.3 匹配策略
 
+- epoch \<= 4，使用 `BatchATSSAssigner`
```
Comment (hhaAndroid) [inline docs/zh_cn/algorithm_descriptions/yolov6_description.md:123]:
```suggestion
- epoch <= 4，使用 `BatchATSSAssigner`
```

Comment (Nioolek) [reply docs/zh_cn/algorithm_descriptions/yolov6_description.md:123]: 说明一下epoch从0开始？

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md [hunk 3]
```diff
@@ -65,26 +65,332 @@ YOLOv6 目标检测算法中使用的数据增强与 YOLOv5 基本一致，唯
...
+- epoch > 4，使用 `BatchTaskAlignedAssigner`
+
+#### ATSSAssigner
+
+ATSSAssigner 是 [ATSS](https://arxiv.org/abs/1912.02424) 中提出的标签匹配策略。
+ATSS的匹配策略简单总结为：**通过中心点距离先验对样本进行初筛,然后自适应生成阈值筛选正样本。**
```
Comment (PeterH0323) [inline docs/zh_cn/algorithm_descriptions/yolov6_description.md:129]:
```suggestion
ATSS 的匹配策略简单总结为：**通过中心点距离先验对样本进行初筛,然后自适应生成阈值筛选正样本。**
```

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md [hunk 4]
```diff
@@ -65,26 +65,332 @@ YOLOv6 目标检测算法中使用的数据增强与 YOLOv5 基本一致，唯
 
 ### 1.3 正负样本匹配策略
 
+YOLOv6 采用的标签匹配策略与 [TOOD](https://arxiv.org/abs/2108.07755)
+相同, 前 4 个 epoch 采用 `ATSSAssigner` 作为标签匹配策略的 `warm-up` ,
+后续使用 `TaskAlignedAssigner` 算法选择正负样本, 基于官方开源代码,  `MMYOLO` 中也对两个 assigner 算法进行了优化, 改进为 `Batch` 维度进行计算，
```
Comment (Nioolek) [inline docs/zh_cn/algorithm_descriptions/yolov6_description.md:70]: 提供下增加效率数值数值会不会更好？

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md [hunk 5]
```diff
@@ -65,26 +65,332 @@ YOLOv6 目标检测算法中使用的数据增强与 YOLOv5 基本一致，唯
...
+
+TaskAlignedAssigner 是 [TOOD](https://arxiv.org/abs/2108.07755) 中提出的一种动态样本匹配策略。
+由于 `ATSSAssigner` 是属于静态标签匹配策略，其选取正样本的策略主要根据 `anchor` 的位置进行挑选，
+并不会随着网络的优化而选取到更好的样本。在目标检测中，分类和回归的任务最终作用于同一个目标，所以
+`TaskAlignedAssigner` 认为样本的选取应该更加关注到对分类以及回归都友好的样本点。
+
+`TaskAlignedAssigner` 的匹配策略简单总结为： **根据分类与回归的分数加权的分数选择正样本**。
+
+1. 对于每一个 `GT`，对所有的 `anchor `基于 **GT类别对应分类分数** 与 **预测框与 GT 的 IoU**
+   的加权得到一个关联分类以及回归的对齐分数 `alignment_metrics`。
```
Comment (PeterH0323) [inline docs/zh_cn/algorithm_descriptions/yolov6_description.md:175]:
```suggestion
1. 对于每一个 `GT`，对所有的 `anchor `基于 **GT类别对应分类分数** 与 **预测框与 GT 的 IoU** 的加权得到一个关联分类以及回归的对齐分数 `alignment_metrics`。
```

Comment (PeterH0323) [reply docs/zh_cn/algorithm_descriptions/yolov6_description.md:175]: 建议回车不要太频繁，有可能导致渲染出错的

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md [hunk 6]
```diff
@@ -65,26 +65,332 @@ YOLOv6 目标检测算法中使用的数据增强与 YOLOv5 基本一致，唯
...
+{math}`p\in[0,1]` 表示分类分数。
+
+1. 对于负样本，即当 {math}`q = 0` 时，标准交叉熵部分为 {math}`-\log(p)`，负样本权重使用 {math}`\alpha p^\gamma` 作为 `focal weight`
+   使样本聚焦与困难样本上，这与 `Focal Loss` 基本一致。
+2. 对于正样本，即当 {math}`q > 0` 时，首先计算标准二值交叉熵部分 {math}`-(qlog(p) +(1-q)log(1-p))`,
+   但是针对正样本的权重设置，`Varifocal Loss` 中并没有采用类似 {math}`\alpha p^\gamma`的方式降权，
+   而是认为在网络的学习过程中正样本相对于负样本的学习信号来说更为重要，所以使用了分类的标签 {math}`q`,
+   即 `IoU` 作为 `focal weight`, 使得聚焦到具有高质量的样本上。
+
+但是 YOLOV6 中的 Varifocal Loss 公式采用 `TOOD` 中的 `Task ALignment Learning (TAL)`,
```
Comment (PeterH0323) [inline docs/zh_cn/algorithm_descriptions/yolov6_description.md:238]:
```suggestion
但是 YOLOv6 中的 Varifocal Loss 公式采用 `TOOD` 中的 `Task ALignment Learning (TAL)`,
```

Comment (PeterH0323) [reply docs/zh_cn/algorithm_descriptions/yolov6_description.md:238]: 其他也检查下

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md [hunk 7]
```diff
@@ -65,26 +65,332 @@ YOLOv6 目标检测算法中使用的数据增强与 YOLOv5 基本一致，唯
...
+   才作为正样本，使得样本能够被 `assign` 到合适的 `FPN` 特征层上。
+
+下图中，(a) 所示中等大小物体被 assign 到 FPN 的中层，(b) 所示偏大的物体被 assign 到 FPN 中检测大物体和偏大物体的两个层。
+
+<div align=center>
+<img src="https://user-images.githubusercontent.com/52028100/204796892-6df4fb85-e2cf-4d84-b85f-d8e8efccc644.png" alt="image"/>
+</div>
+
+```python
+# 1. 计算 anchors 与 GT 的 IoU
```
Comment (Nioolek) [inline docs/zh_cn/algorithm_descriptions/yolov6_description.md:145]: 建议这里优化一下。因为上面说了v6是anchor free，但这里又提到了anchor。建议专门用一小节，或者画张图，来说明下这里的anchor和v5中anchor的区别，v6的anchor实际上是相较于stride而言

Comment (Nioolek) [reply docs/zh_cn/algorithm_descriptions/yolov6_description.md:145]: 也用一句话提一下mmyolo中prior=anchor

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md [hunk 8]
```diff
@@ -65,26 +65,332 @@ YOLOv6 目标检测算法中使用的数据增强与 YOLOv5 基本一致，唯
...
+1. 对于负样本，即当 {math}`q = 0` 时，标准交叉熵部分为 {math}`-\log(p)`，负样本权重使用 {math}`\alpha p^\gamma` 作为 `focal weight`
+   使样本聚焦与困难样本上，这与 `Focal Loss` 基本一致。
+2. 对于正样本，即当 {math}`q > 0` 时，首先计算标准二值交叉熵部分 {math}`-(qlog(p) +(1-q)log(1-p))`,
+   但是针对正样本的权重设置，`Varifocal Loss` 中并没有采用类似 {math}`\alpha p^\gamma`的方式降权，
+   而是认为在网络的学习过程中正样本相对于负样本的学习信号来说更为重要，所以使用了分类的标签 {math}`q`,
+   即 `IoU` 作为 `focal weight`, 使得聚焦到具有高质量的样本上。
+
+但是 YOLOV6 中的 Varifocal Loss 公式采用 `TOOD` 中的 `Task ALignment Learning (TAL)`,
+将预测的 `IoU` 根据之前标签匹配策略中的分类对齐度 `alignment_metrics` 进行了归一化,
+得到归一化 {math}`\hat{t}`。
```
Comment (Nioolek) [inline docs/zh_cn/algorithm_descriptions/yolov6_description.md:240]: 这样子公式在make html后能正常显示吗？

Comment (satuoqaq) [reply docs/zh_cn/algorithm_descriptions/yolov6_description.md:240]: 我不知道，我windows渲染不了，make html会出错

File: docs/zh_cn/algorithm_descriptions/yolov6_description.md (hunk 2 above)
Comment (satuoqaq) [reply docs/zh_cn/algorithm_descriptions/yolov6_description.md:123]: 我写的意思是从第1个开始到4个
或者换成这样写会不会更好
- 0 <= epoch < 4，使用 `BatchATSSAssigner`
- epoch >= 4，使用 `BatchTaskAlignedAssigner`

**Review by hhaAndroid — APPROVED** [review]
